pennylane/gradients/pulse_gradient.py

```diff
@@ -0,0 +1,547 @@
+    return [
+        qml.tape.QuantumScript(ops_pre + split + ops_post, tape.measurements)
+        for split in split_evolve_ops
+    ]
+
+
+def _parshift_and_integrate(results, cjacs, int_prefactor, single_measure, shot_vector):
+    """Apply the parameter-shift rule post-processing to tape results and contract
+    with classical Jacobians, effectively evaluating the numerical integral of the stochastic
+    parameter-shift rule.
+    This is a helper method for the new return type system.
+
+    Args:
+        results (list): Tape evaluation results, corresponding to the modified quantum
+            circuit result when using the applicable parameter shifts and the sample splitting
+            times. Results should be ordered such that the different shifted circuits for a given
+            splitting time are grouped together
+        cjacs (tensor_like): classical Jacobian evaluated at the splitting times
+        int_prefactor (float): prefactor of the numerical integration, corresponding to the size
+            of the time range divided by the number of splitting time samples
+        single_measure (bool):
+        shot_vector (bool):
+    Returns:
+        tensor_like or tuple[tensor_like] or tuple[tuple[tensor_like]]: Gradient entry
+    """
+
+    def _diff_and_contract(res_list, cjacs, int_prefactor):
+        diff = qml.math.stack(res_list)[::2] - qml.math.stack(res_list)[1::2]
+        return qml.math.tensordot(diff, cjacs, axes=[[0], [0]]) * int_prefactor
+
+    # If multiple measure xor shot_vector: One axis to pull out of the shift rule and integration
+    if not single_measure + shot_vector == 1:
+        return tuple(_diff_and_contract(r, cjacs, int_prefactor) for r in zip(*results))
+    if single_measure:
+        # Single measurement without shot vector
+        return _diff_and_contract(results, cjacs, int_prefactor)
+
+    # Multiple measurements with shot vector
+    return tuple(
+        tuple(_diff_and_contract(_r, cjacs, int_prefactor) for _r in zip(*r)) for r in zip(*results)
+    )
+
+
+def _stoch_pulse_grad(tape, argnum=None, num_samples=1, sampler_seed=None, shots=None):
+    r"""Compute the gradient of a quantum circuit composed of pulse sequences by applying the
+    stochastic parameter shift rule. For a pulse-based cost function :math:`C(\boldsymbol{v}, T)`
+    with variational parameters :math:`\boldsymbol{v}` and evolution time :math:`T`, it is given by
+    (c.f. Eqn. (6) in `Leng et al. (2022) <https://arxiv.org/abs/2210.15812>`__ with altered
+    notation):
+
+    .. math::
+
+        \frac{\partial C}{\partial v_k}
+        = \int_{0}^{T} \mathrm{d}\tau \sum_{j=1}^m
+        \frac{\partial f_j}{\partial v_k}(\boldsymbol{v}, \tau)
+        \left[C_j^{(+)}(\boldsymbol, \tau) - C_j^{(-)}(\boldsymbol, \tau)\right]
+
+    Here, :math:`f_j` are the pulse envelopes that capture the time dependence of the pulse
+    Hamiltonian and :math:`C_j^{(\pm)}` are modified cost functions.
```

Comment:
I think you can even expand here and mention the form of $H = H_d + \sum_i f_i(v, t) H_i$ and $C_j^\pm = \langle ... \rangle$

Reply:
:heavy_check_mark: